File: src/ruby/lib/geodelta/packer64.rb
# encoding: utf-8

module GeoDelta
  class Packer64
    def pack_world_delta(id)
      return id << 59
    end

    def unpack_world_delta(value)
      return (value >> 59) & 0b111
    end

    def pack_sub_delta(level, id)
      return id << (57 - ((level - 2) * 2))
    end

    def unpack_sub_delta(level, value)
      return (value >> (57 - ((level - 2) * 2))) & 0b11
    end

    def pack_level(level)
      return level
    end

    def unpack_level(level)
      return level & 0b11111
    end

    def pack(ids)
      wid   = self.pack_world_delta(ids[0])
      sids  = ids[1..-1].each_with_index.map { |id, i| self.pack_sub_delta(i + 2, id) }.inject(0, &:+)
      level = self.pack_level(ids.size)
      return wid + sids + level
    end

    def unpack(value)
      level = self.unpack_level(value)
      wid   = self.unpack_world_delta(value)
      sids  = (level - 1).times.map { |i| self.unpack_sub_delta(i + 2, value) }
      return [wid] + sids
    end
  end
end

File: src/java/geodelta/src/jp/nayutaya/geodelta/Packer64.java

package jp.nayutaya.geodelta;

/**
 * 64ビット符号付き整数にパックされたデルタID列を扱うクラス。
 */
public class Packer64
{
    /**
     * ワールドデルタIDをパックする。
     *
     * @param id ワールドデルタID
     * @return パックされたID
     */
    /* package */static long packWorldDelta(final byte id)
    {
        return (long)id << 59;
    }

    /**
     * ワールドデルタIDをアンパックする。
     *
     * @param value パックされたデルタID列
     * @return ワールドデルタID
     */
    /* package */static byte unpackWorldDelta(final long value)
    {
        return (byte)((value >> 59) & 0x07);
    }

    /**
     * サブデルタIDをパックする。
     *
     * @param level サブデルタのレベル
     * @param id サブデルタID
     * @return パックされたID
     */
    /* package */static long packSubDelta(final int level, final byte id)
    {
        return (long)id << (57 - ((level - 2) * 2));
    }

    /**
     * サブデルタIDをアンパックする。
     *
     * @param level サブデルタのレベル
     * @param value パックされたデルタID列
     * @return サブデルタID
     */
    /* package */static byte unpackSubDelta(final int level, final long value)
    {
        return (byte)((value >> (57 - ((level - 2) * 2))) & 0x03);
    }

    /**
     * レベルをパックする。
     *
     * @param level レベル
     * @return パックされたID
     */
    /* package */static long packLevel(final int level)
    {
        return level;
    }

    /**
     * レベルをアンパックする。
     *
     * @param value パックされたデルタID列
     * @return レベル
     */
    /* package */static int unpackLevel(final long value)
    {
        return (int)(value & 0x1F);
    }

    /**
     * デルタID列を64ビット整数にパックする。
     *
     * @param ids デルタID列
     * @return パックされたデルタID列
     */
    public static long pack(final byte[] ids)
    {
        final long wid = packWorldDelta(ids[0]);
        long sid = 0;
        for ( int i = 1, len = ids.length; i < len; i++ )
        {
            sid |= packSubDelta(i + 1, ids[i]);
        }
        final long level = packLevel(ids.length);
        return wid | sid | level;
    }

    /**
     * 64ビット整数からデルタID列をアンパックする。
     *
     * @param value パックされたデルタID列
     * @return デルタID列
     */
    public static byte[] unpack(final long value)
    {
        final int level = unpackLevel(value);
        final byte[] ids = new byte[level];
        ids[0] = unpackWorldDelta(value);
        for ( int i = 1; i < level; i++ )
        {
            ids[i] = unpackSubDelta(i + 1, value);
        }
        return ids;
    }
}
